src/OpenTelemetry/Metrics/Tags.cs

```diff
@@ -52,33 +34,102 @@ public Tags(KeyValuePair<string, object?>[] keyValuePairs)
 
     public readonly bool Equals(Tags other)
     {
-        var length = this.KeyValuePairs.Length;
+        var ourKvps = this.KeyValuePairs;
+        var theirKvps = other.KeyValuePairs;
 
-        if (length != other.KeyValuePairs.Length)
+        if (ourKvps.Length != theirKvps.Length)
         {
             return false;
         }
 
-        for (int i = 0; i < length; i++)
+#if NET6_0_OR_GREATER
+        // Note: This loop uses unsafe code (pointers) to elide bounds checks on
+        // two arrays we know to be of equal length.
+        var cursor = ourKvps.Length;
+        if (cursor > 0)
+        {
+            ref var ours = ref MemoryMarshal.GetArrayDataReference(ourKvps);
+            ref var theirs = ref MemoryMarshal.GetArrayDataReference(theirKvps);
+            while (true)
+            {
+                // Equality check for Keys
+                if (!ours.Key.Equals(theirs.Key))
+                {
+                    return false;
+                }
+
+                // Equality check for Values
+                if (!ours.Value?.Equals(theirs.Value) ?? theirs.Value != null)
+                {
+                    return false;
+                }
+
+                if (--cursor == 0)
+                {
+                    break;
+                }
+
+                ours = ref Unsafe.Add(ref ours, 1);
+                theirs = ref Unsafe.Add(ref theirs, 1);
+            }
+        }
+#else
+        for (int i = 0; i < ourKvps.Length; i++)
         {
-            ref var left = ref this.KeyValuePairs[i];
-            ref var right = ref other.KeyValuePairs[i];
+            ref var ours = ref ourKvps[i];
+
+            // Note: Bounds check happens here for theirKvps element access
+            ref var theirs = ref theirKvps[i];
 
             // Equality check for Keys
-            if (!left.Key.Equals(right.Key, StringComparison.Ordinal))
+            if (!ours.Key.Equals(theirs.Key))
```

Comment:
mm.. is removing Ordinal comparison *faster* ?

Reply:
It isn't removing ordinal comparison so much as it is getting to it faster 🤣

What I see before is call to `string.Equals(string, int)`:

```asm
call     [System.String:Equals(System.String,int):bool:this]
```

What I see after is inlined call to `System.SpanHelpers:SequenceEqual`:

```asm
call     [System.SpanHelpers:SequenceEqual(byref,byref,ulong):bool]
```

* [string.Equals(string) code](https://github.com/dotnet/runtime/blob/aea36820fdef48d6ef375508977e2f681fa4e199/src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs#L612)

* [string.Equals(string int) code](https://github.com/dotnet/runtime/blob/aea36820fdef48d6ef375508977e2f681fa4e199/src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs#L647C48-L647C94)

For ordinal both eventually call into [EqualsHelper](https://github.com/dotnet/runtime/blob/aea36820fdef48d6ef375508977e2f681fa4e199/src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs#L22) which is where the `SpanHelpers.SequenceEqual` comes into play.

Reply:
> For ordinal both eventually call into EqualsHelper which is where the SpanHelpers.SequenceEqual comes into play.

Are you saying that the default behavior for `string.Equals(string obj)` is to do an `Ordinal` comparison anyway? If yes, then maybe add a comment that we are doing an ordinal comparison but not explicitly specifying it to inline calls and make it faster.

Reply:
interesting.. do we want to rely on .NET's implementation detail?
Or maybe the question I really wanted to ask was "Isn't `StringComparison.Ordinal` what we want to be doing? (assume no perf diff.) "

Reply:
Yes `string.Equals(string)` does an ordinal comparison! Check out the remarks on: https://learn.microsoft.com/dotnet/api/system.string.equals?#system-string-equals(system-string)

I'll add a note on there.

Reply:
> For ordinal both eventually call into EqualsHelper which is where the SpanHelpers.SequenceEqual comes into play.

I found this to be the case for `net6.0` and higher but it looks like for `net462` the calls to `string.Equals(string value)` and `string.Equals(string value, StringComparison.Ordinal)` lead to different code paths with the latter one having a special (possibly optimized) path for ASCII strings. Maybe remove the `StringComparison.Ordinal` parameter only for `NET6_0_OR_GREATER`.

Reply:
@utpilla Interesting! Where do you see that? I just went poking around and it looks pretty similar on net framework: https://github.com/microsoft/referencesource/blob/51cf7850defa8a17d815b4700b67116e3fa283c2/mscorlib/system/string.cs#L614-L617

I see the special ascii code in the `StringComparison.OrdinalIgnoreCase` case is that what you were referring to?

Reply:
> I see the special ascii code in the StringComparison.OrdinalIgnoreCase case is that what you were referring to?

Ah yes. I was looking at the `OrdinalIgnoreCase` 😄